### DISCUSSION TITLE
Performance Consideration

### DISCUSSION THREAD

**Discussion by infinitydon** [nextmn/.github#4] [Q&A] [answered]
Hi,

Just want to ask if it is possible to consider using VPP instead, using VPP has the added benefit of srv6 MUP (https://docs.fd.io/vpp/23.06/developer/plugins/srv6/mobile_plugin_doc.html?highlight=srv6) plus user plane acceleration as part of it's dataplane features. The only missing piece will be the control-plane that will be translating the PFCP to srv6 configuration in VPP.

Also do you think the **_tun_** interface is performant? Some throughput tests for Open5gs UPF **_tun_** interface could not do more than 1Gbps

**Comment by louisroyer** [comment, marked as answer]
> Just want to ask if it is possible to consider using VPP instead,

As far as I understand, it would be difficult (or even not doable with [current GTPU plugin](https://s3-docs.fd.io/vpp/23.10/cli-reference/clis/clicmd_src_plugins_gtpu.html)) to translate advanced rules such as SDF Flows.

>  using VPP has the added benefit of srv6 MUP (https://docs.fd.io/vpp/23.06/developer/plugins/srv6/mobile_plugin_doc.html?highlight=srv6) plus user plane acceleration as part of it's dataplane features. The only missing piece will be the control-plane that will be translating the PFCP to srv6 configuration in VPP.

I already started to integrate SRv6 MUP inside my Free5gc based testbed, and considered using VPP to create endpoints, but since all is not implemented with enough control on SR policies for my usecases (especially `H.M.GTP4.D`, which I need with my testbed, and `End.M.GTP6.D` which doesn't allow to use complex SR Policies), I started implementing SRv6 MUP myself on [nextmn/srv6](github.com/nextmn/srv6) (still a work in progress). However, VPP with the SRv6 plugin may be an alternative backend for some Endpoints Behaviors in the future. This will not solve entirely perf issues since we will still have an unoptimized function for the SRGW on uplink. 

> Also do you think the tun interface is performant? Some throughput tests for Open5gs UPF tun interface could not do more than 1Gbps

From what I read about tun interfaces, [when UERANSIM's dev had tested it](https://github.com/aligungr/UERANSIM/issues/370#issuecomment-892445153), he concluded it depends highly on the test environment.

Maybe using [gopacket's RawConn](https://pkg.go.dev/golang.org/x/net/ipv4#RawConn) would have better performances…

Anyway, currently on nextmn/upf my bottleneck is on applying PFCP Rules, and I don't even reach 1Gbps. This implementation would never beat a Linux module anyway: the main goal was only to have an UPF that is easy to modify and to install related to [a kernel module](https://github.com/free5gc/gtp5g).

**Comment by infinitydon** [comment]
Thanks @louisroyer, will keep an eye on the updates, project looks very interesting.
